src/sql.rs

```diff
@@ -38,20 +39,55 @@ mod migrations;
 /// A wrapper around the underlying Sqlite3 object.
 #[derive(Debug)]
 pub struct Sql {
+    /// Database file path
+    pub(crate) dbfile: PathBuf,
+
     pool: RwLock<Option<r2d2::Pool<r2d2_sqlite::SqliteConnectionManager>>>,
+
+    /// SQLCipher passphrase.
+    ///
+    /// Empty string if database is not encrypted.
+    passphrase: RwLock<String>,
 }
 
-impl Default for Sql {
-    fn default() -> Self {
+impl Sql {
+    pub fn new(dbfile: PathBuf) -> Sql {
         Self {
-            pool: RwLock::new(None),
+            dbfile,
+            pool: Default::default(),
+            passphrase: Default::default(),
         }
     }
-}
 
-impl Sql {
-    pub fn new() -> Sql {
-        Self::default()
+    /// Sets SQLCipher passphrase for key derivation.
+    ///
+    /// Returns true if passphrase is correct, i.e. the database is new or can be unlocked with
+    /// this passphrase, and false if the database is already encrypted with another passphrase or
+    /// corrupted.
+    ///
+    /// Fails if database is already open.
+    pub async fn set_passphrase(&self, passphrase: String) -> Result<bool> {
+        if self.is_open().await {
+            bail!("Database is already opened.");
+        }
+
+        // Hold the lock to prevent other thread from opening the database.
+        let _lock = self.pool.write().await;
+
+        // Test that the key is correct using a single connection.
+        let connection = Connection::open(&self.dbfile)?;
+        connection
+            .pragma_update(None, "key", &passphrase)
+            .context("failed to set PRAGMA key")?;
+        let key_is_correct = connection
+            .query_row("SELECT count(*) FROM sqlite_master", [], |_row| Ok(()))
```

Comment:
@link2xt why doesn't this end with a ";"? Is this a bug? If so, why doesn't opening an encrypted database always fail?

Reply:
`;` is not needed, no idea why we put it everywhere.

Reply:
Neither Rusqlite nor sqlx require it. If you look at examples at https://lib.rs/crates/rusqlite, they don't have `;` at the end.